library/Xi/Sms/Gateway/BulkSmsGateway.php

```diff
@@ -0,0 +1,292 @@
+        $url = $this->endpoint;
+
+        foreach ($message->getTo() as $to) {
+
+            switch($this->bits) {
+                case 7:
+                    //
+                    // Sending 7-bit message
+                    //
+                    $post_body = $this->seven_bit_sms($body, $to);
+                    $this->getClient()->post($url, array(), $post_body);
+                break;
+                
+                case 16:
+                    //
+                    // Sending unicode message
+                    //
+                    $post_body = $this->unicode_sms($body, $to);
+                    $this->getClient()->post($url, array(), $post_body);
+                break;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Make the post body for a 7 bit SMS message.
+     *
+     * @param string $message
+     * @param string $msisdn
+     * @return string
+     */
+    protected function seven_bit_sms($message, $msisdn) {
+        $post_fields = array (
+            'username' => $this->username,
+            'password' => $this->password,
+            'message'  => $this->character_resolve( $message ),
+            'msisdn'   => $msisdn
+        );
+
+        return $this->make_post_body($post_fields);
+    }
+
+    /**
+     * Make the post body for a unicode SMS message.
+     *
+     * @param string $username
+     * @param string $password
+     * @param string $message
+     * @param string $msisdn
+     * @return string
+     */
+    protected function unicode_sms($message, $msisdn) {
+        $post_fields = array (
+            'username' => $this->username,
+            'password' => $this->password,
+            'message'  => $this->string_to_utf16_hex($message),
+            'msisdn'   => $msisdn,
+            'dca'      => '16bit'
+        );
+
+        return make_post_body($post_fields);
+    }
+
+    /**
+     * Resolve 7 bit characters
+     *
+     * @param string $body
+     * @return string
+     */
+    protected function character_resolve($body) {
+        $special_chrs = array(
+            'Î”'=>'0xD0', 'Î¦'=>'0xDE', 'Î“'=>'0xAC', 'Î›'=>'0xC2', 'Î©'=>'0xDB',
+            'Î '=>'0xBA', 'Î¨'=>'0xDD', 'Î£'=>'0xCA', 'Î˜'=>'0xD4', 'Îž'=>'0xB1',
+            'Â¡'=>'0xA1', 'Â£'=>'0xA3', 'Â¤'=>'0xA4', 'Â¥'=>'0xA5', 'Â§'=>'0xA7',
+            'Â¿'=>'0xBF', 'Ã„'=>'0xC4', 'Ã…'=>'0xC5', 'Ã†'=>'0xC6', 'Ã‡'=>'0xC7',
+            'Ã‰'=>'0xC9', 'Ã‘'=>'0xD1', 'Ã–'=>'0xD6', 'Ã˜'=>'0xD8', 'Ãœ'=>'0xDC',
+            'ÃŸ'=>'0xDF', 'Ã '=>'0xE0', 'Ã¤'=>'0xE4', 'Ã¥'=>'0xE5', 'Ã¦'=>'0xE6',
+            'Ã¨'=>'0xE8', 'Ã©'=>'0xE9', 'Ã¬'=>'0xEC', 'Ã±'=>'0xF1', 'Ã²'=>'0xF2',
+            'Ã¶'=>'0xF6', 'Ã¸'=>'0xF8', 'Ã¹'=>'0xF9', 'Ã¼'=>'0xFC',
+        );
+
+        $ret_msg = '';
+        if( mb_detect_encoding($body, 'UTF-8') != 'UTF-8' ) {
+                        $body = utf8_encode($body);
+                }
+                for ( $i = 0; $i < mb_strlen( $body, 'UTF-8' ); $i++ ) {
```

Comment:
Would the following work?

    $from = array_keys($special_chrs);
    $to = array_values($special_chrs);
    $ret_msg = str_replace($from, $to, $body);